sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,442 @@
+
+    val memMode = OrcColumnarBatchReader.DEFAULT_MEMORY_MODE
+    val capacity = ColumnarBatch.DEFAULT_BATCH_SIZE
+    if (memMode == MemoryMode.OFF_HEAP) {
+      columnVectors = OffHeapColumnVector.allocateColumns(capacity, resultSchema)
+    } else {
+      columnVectors = OnHeapColumnVector.allocateColumns(capacity, resultSchema)
+    }
+    columnarBatch = new ColumnarBatch(resultSchema, columnVectors.toArray, capacity)
+
+    if (partitionValues.numFields > 0) {
+      val partitionIdx = requiredSchema.fields.length
+      for (i <- 0 until partitionValues.numFields) {
+        ColumnVectorUtils.populate(columnVectors(i + partitionIdx), partitionValues, i)
+        columnVectors(i + partitionIdx).setIsConstant()
+      }
+    }
+  }
+
+  /**
+   * Return true if there exists more data in the next batch. If exists, prepare the next batch
+   * by copying from ORC VectorizedRowBatch columns to Spark ColumnarBatch columns.
+   */
+  private def nextBatch(): Boolean = {
+    if (rowsReturned >= totalRowCount) {
+      return false
+    }
+
+    rows.nextBatch(batch)
+    val batchSize = batch.size
+    if (batchSize == 0) {
+      return false
+    }
+    rowsReturned += batchSize
+    columnarBatch.reset()
+    columnarBatch.setNumRows(batchSize)
+
+    var i = 0
+    while (i < requiredSchema.length) {
+      val field = requiredSchema(i)
+      val toColumn = columnVectors(i)
+
+      if (requestedColIds(i) < 0) {
+        toColumn.appendNulls(batchSize)
+      } else {
+        val fromColumn = batch.cols(requestedColIds(i))
+
+        if (fromColumn.isRepeating) {
+          if (fromColumn.isNull(0)) {
+            toColumn.appendNulls(batchSize)
+          } else {
+            field.dataType match {
+              case BooleanType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0) == 1
+                toColumn.appendBooleans(batchSize, data)
+
+              case ByteType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toByte
+                toColumn.appendBytes(batchSize, data)
+              case ShortType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toShort
+                toColumn.appendShorts(batchSize, data)
+              case IntegerType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+              case LongType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0)
+                toColumn.appendLongs(batchSize, data)
+
+              case DateType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+
+              case TimestampType =>
+                val data = fromColumn.asInstanceOf[TimestampColumnVector]
+                toColumn.appendLongs(batchSize, data.time(0) * 1000L + data.nanos(0) / 1000L)
```

Comment:
This raw calculation here should be encapsulated in a utilities method with a few comments explaining what is happening (i.e. the conversion from TimestampColumnVector's java.sql.Timestamp representation to Spark's).  It is repeated (so to speak several times in this method/